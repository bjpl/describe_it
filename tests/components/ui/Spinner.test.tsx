import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, it, expect } from 'vitest';
import '@testing-library/jest-dom';
import {
  LoadingOverlay,
  PageLoader,
  ButtonLoader,
  InlineLoader,
} from '@/components/ui/LoadingStates';

describe('Loading Components', () => {
  describe('LoadingOverlay', () => {
    it('should render children', () => {
      render(
        <LoadingOverlay isLoading={false}>
          <div>Content</div>
        </LoadingOverlay>
      );
      expect(screen.getByText('Content')).toBeInTheDocument();
    });

    it('should show overlay when loading', () => {
      const { container } = render(
        <LoadingOverlay isLoading={true}>
          <div>Content</div>
        </LoadingOverlay>
      );
      expect(container.querySelector('.absolute.inset-0')).toBeInTheDocument();
    });

    it('should not show overlay when not loading', () => {
      const { container } = render(
        <LoadingOverlay isLoading={false}>
          <div>Content</div>
        </LoadingOverlay>
      );
      expect(container.querySelector('.absolute.inset-0')).not.toBeInTheDocument();
    });

    it('should display custom loading message', () => {
      render(
        <LoadingOverlay isLoading={true} message="Please wait...">
          <div>Content</div>
        </LoadingOverlay>
      );
      expect(screen.getByText('Please wait...')).toBeInTheDocument();
    });

    it('should apply custom className', () => {
      const { container } = render(
        <LoadingOverlay isLoading={false} className="custom-overlay">
          <div>Content</div>
        </LoadingOverlay>
      );
      expect(container.firstChild).toHaveClass('custom-overlay');
    });

    it('should have backdrop blur when loading', () => {
      const { container } = render(
        <LoadingOverlay isLoading={true}>
          <div>Content</div>
        </LoadingOverlay>
      );
      expect(container.querySelector('.backdrop-blur-sm')).toBeInTheDocument();
    });
  });

  describe('PageLoader', () => {
    it('should render with default message', () => {
      render(<PageLoader />);
      expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('should render with custom message', () => {
      render(<PageLoader message="Loading data..." />);
      expect(screen.getByText('Loading data...')).toBeInTheDocument();
    });

    it('should have full screen height', () => {
      const { container } = render(<PageLoader />);
      expect(container.firstChild).toHaveClass('min-h-screen');
    });

    it('should center content', () => {
      const { container } = render(<PageLoader />);
      expect(container.firstChild).toHaveClass('flex', 'items-center', 'justify-center');
    });

    it('should apply custom className', () => {
      const { container } = render(<PageLoader className="custom-page-loader" />);
      expect(container.firstChild).toHaveClass('custom-page-loader');
    });
  });

  describe('ButtonLoader', () => {
    it('should render children when not loading', () => {
      render(<ButtonLoader loading={false}>Submit</ButtonLoader>);
      expect(screen.getByText('Submit')).toBeInTheDocument();
    });

    it('should show spinner when loading', () => {
      const { container } = render(<ButtonLoader loading={true}>Submit</ButtonLoader>);
      const spinner = container.querySelector('.animate-spin');
      expect(spinner).toBeInTheDocument();
    });

    it('should display loading text when provided', () => {
      render(<ButtonLoader loading={true} loadingText="Submitting...">Submit</ButtonLoader>);
      expect(screen.getByText('Submitting...')).toBeInTheDocument();
    });

    it('should display original children as loading text if no loadingText provided', () => {
      render(<ButtonLoader loading={true}>Submit</ButtonLoader>);
      expect(screen.getByText('Submit')).toBeInTheDocument();
    });

    it('should render small spinner', () => {
      const { container } = render(<ButtonLoader loading={true} size="sm">Save</ButtonLoader>);
      const spinner = container.querySelector('.h-3.w-3');
      expect(spinner).toBeInTheDocument();
    });

    it('should render medium spinner', () => {
      const { container } = render(<ButtonLoader loading={true} size="md">Save</ButtonLoader>);
      const spinner = container.querySelector('.h-4.w-4');
      expect(spinner).toBeInTheDocument();
    });

    it('should render large spinner', () => {
      const { container } = render(<ButtonLoader loading={true} size="lg">Save</ButtonLoader>);
      const spinner = container.querySelector('.h-5.w-5');
      expect(spinner).toBeInTheDocument();
    });

    it('should have aria-hidden on spinner', () => {
      const { container } = render(<ButtonLoader loading={true}>Save</ButtonLoader>);
      const spinner = container.querySelector('[aria-hidden="true"]');
      expect(spinner).toBeInTheDocument();
    });
  });

  describe('InlineLoader', () => {
    it('should render spinner', () => {
      const { container } = render(<InlineLoader />);
      const spinner = container.querySelector('.animate-spin');
      expect(spinner).toBeInTheDocument();
    });

    it('should render extra small size', () => {
      const { container } = render(<InlineLoader size="xs" />);
      const spinner = container.querySelector('.h-3.w-3');
      expect(spinner).toBeInTheDocument();
    });

    it('should render small size', () => {
      const { container } = render(<InlineLoader size="sm" />);
      const spinner = container.querySelector('.h-4.w-4');
      expect(spinner).toBeInTheDocument();
    });

    it('should render medium size', () => {
      const { container } = render(<InlineLoader size="md" />);
      const spinner = container.querySelector('.h-5.w-5');
      expect(spinner).toBeInTheDocument();
    });

    it('should apply custom className', () => {
      const { container } = render(<InlineLoader className="custom-spinner" />);
      expect(container.querySelector('.custom-spinner')).toBeInTheDocument();
    });

    it('should have aria-label', () => {
      const { container } = render(<InlineLoader />);
      const spinner = container.querySelector('[aria-label="Loading"]');
      expect(spinner).toBeInTheDocument();
    });

    it('should render SVG element', () => {
      const { container } = render(<InlineLoader />);
      const svg = container.querySelector('svg');
      expect(svg).toBeInTheDocument();
    });

    it('should have correct SVG attributes', () => {
      const { container } = render(<InlineLoader />);
      const svg = container.querySelector('svg');
      expect(svg).toHaveAttribute('fill', 'none');
      expect(svg).toHaveAttribute('viewBox', '0 0 24 24');
    });
  });

  describe('Spinner Animation', () => {
    it('should have animate-spin class', () => {
      const { container } = render(<InlineLoader />);
      expect(container.querySelector('.animate-spin')).toBeInTheDocument();
    });

    it('should have circular path', () => {
      const { container } = render(<InlineLoader />);
      const circle = container.querySelector('circle');
      expect(circle).toBeInTheDocument();
    });

    it('should have progress path', () => {
      const { container } = render(<InlineLoader />);
      const path = container.querySelector('path');
      expect(path).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('should have proper ARIA attributes for inline loader', () => {
      const { container } = render(<InlineLoader />);
      const spinner = container.querySelector('svg');
      expect(spinner).toHaveAttribute('aria-label', 'Loading');
    });

    it('should have role region for page loader', () => {
      const { container } = render(<PageLoader />);
      // PageLoader renders a div container
      expect(container.firstChild).toBeInTheDocument();
    });
  });

  describe('Snapshot Tests', () => {
    it('should match snapshot for LoadingOverlay', () => {
      const { container } = render(
        <LoadingOverlay isLoading={true}>
          <div>Content</div>
        </LoadingOverlay>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for PageLoader', () => {
      const { container } = render(<PageLoader message="Loading..." />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for ButtonLoader', () => {
      const { container } = render(<ButtonLoader loading={true}>Save</ButtonLoader>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for InlineLoader', () => {
      const { container } = render(<InlineLoader />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
