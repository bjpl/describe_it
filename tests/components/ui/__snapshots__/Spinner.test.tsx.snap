// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loading Components > Snapshot Tests > should match snapshot for ButtonLoader 1`] = `
<svg
  aria-hidden="true"
  class="animate-spin mr-2 h-4 w-4"
  fill="none"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    class="opacity-25"
    cx="12"
    cy="12"
    r="10"
    stroke="currentColor"
    stroke-width="4"
  />
  <path
    class="opacity-75"
    d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
    fill="currentColor"
  />
</svg>
`;

exports[`Loading Components > Snapshot Tests > should match snapshot for InlineLoader 1`] = `
<svg
  aria-label="Loading"
  class="animate-spin h-4 w-4"
  fill="none"
  viewBox="0 0 24 24"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    class="opacity-25"
    cx="12"
    cy="12"
    r="10"
    stroke="currentColor"
    stroke-width="4"
  />
  <path
    class="opacity-75"
    d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
    fill="currentColor"
  />
</svg>
`;

exports[`Loading Components > Snapshot Tests > should match snapshot for LoadingOverlay 1`] = `
<div
  class="relative"
>
  <div>
    Content
  </div>
  <div
    class="absolute inset-0 bg-background/80 backdrop-blur-sm flex items-center justify-center z-50"
  >
    <div
      class="text-center"
    >
      <div
        class="flex flex-col items-center justify-center space-y-3 "
      >
        <div
          class="flex items-center justify-center"
        >
          <svg
            class="lucide lucide-loader2 w-6 h-6 animate-spin"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 12a9 9 0 1 1-6.219-8.56"
            />
          </svg>
        </div>
        <p
          class="text-center text-gray-600 text-base"
        >
          Loading...
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Loading Components > Snapshot Tests > should match snapshot for PageLoader 1`] = `
<div
  class="min-h-screen flex items-center justify-center"
>
  <div
    class="flex flex-col items-center justify-center space-y-3 "
  >
    <div
      class="flex items-center justify-center"
    >
      <svg
        class="lucide lucide-loader2 w-12 h-12 animate-spin"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 12a9 9 0 1 1-6.219-8.56"
        />
      </svg>
    </div>
    <p
      class="text-center text-gray-600 text-lg"
    >
      Loading...
    </p>
  </div>
</div>
`;
